nuxeo-core/nuxeo-core-io/src/main/java/org/nuxeo/ecm/core/io/marshallers/json/enrichers/BasePermissionsJsonEnricher.java

```diff
@@ -85,14 +90,13 @@ public void write(JsonGenerator jg, DocumentModel document) throws IOException {
         jg.writeEndArray();
     }
 
-    private Iterable<String> getPermissionsInSession(final DocumentModel doc, final CoreSession session) {
-        final Principal principal = session.getPrincipal();
-        return Iterables.filter(availablePermissions, new Predicate<String>() {
-            @Override
-            public boolean apply(String permission) {
-                return session.hasPermission(principal, doc.getRef(), permission);
-            }
-        });
+    private List<String> getPermissionsInSession(DocumentModel doc, CoreSession session) {
+        Principal principal = session.getPrincipal();
+        PermissionProvider permissionProvider = Framework.getService(PermissionProvider.class);
+        Set<String> permissions = permissionProvider.getUserVisiblePermissionDescriptors(doc.getType()).stream().map(
+            UserVisiblePermission::getId).collect(Collectors.toSet());
+        permissions.addAll(availablePermissions);
+        return session.filterGrantedPermissions(principal, doc.getRef(), new ArrayList<>(permissions));
```

Comment:
You can use `List` instead of `Set` when creating the permission. In this case, there's no need to change it at the end.

```java
List<String> permissions = ...
        .stream()
        .map(UserVisiblePermission::getId)
        .collect(Collectors.toList());
```